nixos/lib/make-iso9660-image.sh

```diff
@@ -29,39 +31,113 @@ addPath() {
 
 stripSlash "$bootImage"; bootImage="$res"
 
-
-if test -n "$bootable"; then
-
-    # The -boot-info-table option modifies the $bootImage file, so
-    # find it in `contents' and make a copy of it (since the original
-    # is read-only in the Nix store...).
-    for ((i = 0; i < ${#targets_[@]}; i++)); do
-        stripSlash "${targets_[$i]}"
-        if test "$res" = "$bootImage"; then
-            echo "copying the boot image ${sources_[$i]}"
-            cp "${sources_[$i]}" boot.img
-            chmod u+w boot.img
-            sources_[$i]=boot.img
-        fi
-    done
-
-    isoBootFlags="-eltorito-boot ${bootImage}
-                  -eltorito-catalog .boot.cat
-                  -no-emul-boot -boot-load-size 4 -boot-info-table
-                  --sort-weight 1 /isolinux" # Make sure isolinux is near the beginning of the ISO
-fi
-
-if test -n "$usbBootable"; then
-    usbBootFlags="-isohybrid-mbr ${isohybridMbrImage}"
+TMP=$(mktemp -d)
+
+mkdir -p append/boot/grub
+mkdir append/EFI
+touch append/boot/grub/device.map
+addPath "/" "$PWD/append"
+
+if test -n "$mbrBootable"; then
+  "$grubMbr/bin/grub-mkstandalone" \
+      --format=i386-pc \
+      --output="$TMP/core.img" \
+      --install-modules="linux normal iso9660 biosdisk memdisk search tar ls configfile" \
+      --modules="linux normal iso9660 biosdisk search" \
+      --locales="" \
+      --fonts="" \
+      "boot/grub/grub.cfg=$grubCfg"
+
+  # TODO: fix this hack to REALLY skip fsprobe
+  "$grubMbr/bin/grub-install" \
+      --target=i386-pc \
+      --boot-directory=$PWD/append/boot \
+      --no-bootsector \
+      --skip-fs-probe \
+      "/dev/null" -v --force || true
+
+  cat \
+      "$grubMbr/lib/grub/i386-pc/cdboot.img" \
+      "$TMP/core.img" \
+  > "$TMP/bios.img"
+
+  addPath /boot/grub/bios.img "$TMP/bios.img"
+
+  isoBootFlags=" -eltorito-boot
+                    boot/grub/bios.img
+                    -no-emul-boot
+                    -boot-load-size 4
+                    -boot-info-table
+                    --eltorito-catalog boot/grub/boot.cat
+                --grub2-boot-info
+                --grub2-mbr $grubMbr/lib/grub/i386-pc/boot_hybrid.img"
 fi
 
 if test -n "$efiBootable"; then
-    efiBootFlags="-eltorito-alt-boot
-                  -e $efiBootImage
-                  -no-emul-boot
-                  -isohybrid-gpt-basdat"
+  "$grubEfi/bin/grub-mkstandalone" \
+      --format=x86_64-efi \
+      --output="$TMP/bootx64.efi" \
+      --locales="" \
+      --fonts="" \
+      "boot/grub/grub.cfg=$grubCfg"
+
+  # TODO: fix this hack to REALLY skip fsprobe
+  "$grubEfi/bin/grub-install" \
+      --target=x86_64-efi \
+      --efi-directory=$PWD/append/EFI \
+      --boot-directory=$PWD/append/boot \
+      --no-nvram \
+      --skip-fs-probe \
```

Comment:
same as above

Reply:
don't know if this is needed, since all modules are in the efi image (added some code to make detection dynamic)
but dunno if later other archs might have similar problems as MBR with image size limit.... I guess no, we'll see later then